src/Npgsql/NpgsqlTypes/NpgsqlCube.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+#pragma warning disable 1591
+
+// ReSharper disable once CheckNamespace
+namespace NpgsqlTypes
+{
+    public struct NpgsqlCube : IEquatable<NpgsqlCube>
+    {
+        static readonly Regex Regex =
+            new(@"^(\[(?=\(.+\)\,\())?((\()?(-?\d+\.?\d*?(?:,-?\d+(?:\.?\d*))*)(?(3)\)|)(,\((-?\d+\.?\d*(?:,-?\d+\.?\d*)*)\))?)(?(1)\]|)$");
+        
+        readonly List<double> _coords;
+        readonly bool _point;
+
+        public NpgsqlCube(IEnumerable<double> lowerLeft, IEnumerable<double> upperRight) : this(lowerLeft.ToList(), upperRight.ToList()) { }
+        
+        NpgsqlCube(List<double> lowerLeft, List<double> upperRight) 
+            : this(lowerLeft.Concat(upperRight).ToList(), upperRight.Count == 0 || lowerLeft.SequenceEqual(upperRight))
+        { }
+        
+        public NpgsqlCube(params double[] coords) : this(coords.ToList(), true) { }
+        
+        public NpgsqlCube(IEnumerable<double> coords, bool point = false) : this(coords.ToList(), point) { }
+
+        NpgsqlCube(List<double> coords, bool point)
+        {
+            if (!point && coords.Count % 2 != 0)
+            {
+                throw new ArgumentException("Not a valid cube: Different point dimensions.", nameof(coords));
+            }
+            
+            _coords = coords;
+            _point = point;
+        }
+
+        /// <summary>
+        /// The number of dimensions in the cube.
+        /// </summary>
+        public int Dimensions => Point ? Coordinates.Count() : Coordinates.Count() / 2;
+
+        /// <summary>
+        /// Is the cube a point.
+        /// </summary>
+        public bool Point => _point;
+
+        /// <summary>
+        /// The coordinates in the cube.
+        /// </summary>
+        public IEnumerable<double> Coordinates => _coords;
+
+        /// <summary>
+        /// The lower left coordinates of the cube.
+        /// </summary>
+        public IEnumerable<double> LowerLeft => _coords.Take(Dimensions);
```

Comment:
Using LINQ to split the list into two each time seems a bit inefficient. Would it be more appropriate to just have two separate `List<double>` for LowerLeft/UpperRight, and split/join when reading/writing in the handler? The two could point to the same list instance for point.

Reply:
That's probably the correct approach, I threw this together quickly so I tried to match the way it was done in the PostgreSQL impl. Now that I think about it, that is a much more efficient means to store the data in .NET.